src/SavingUsdsTokenWrapper.sol

```diff
@@ -0,0 +1,108 @@
+// SPDX-License-Identifier: AGPL-3.0
+pragma solidity ^0.8.10;
+import 'forge-std/console2.sol';
+
+import {IPool} from 'aave-v3-core/contracts/interfaces/IPool.sol';
+import {ICreditDelegationToken} from 'aave-v3-core/contracts/interfaces/ICreditDelegationToken.sol';
+import {IERC20} from 'aave-v3-core/contracts/dependencies/openzeppelin/contracts/IERC20.sol';
+import {IUSDS} from './dependencies/IUSDS.sol';
+import {BaseTokenWrapper} from './BaseTokenWrapper.sol';
+
+/**
+ * @title SavingUsdsTokenWrapper
+ * @author Aave
+ * @notice Contract to wrap USDS to SuSDS on supply to Aave, or unwrap from SuSDS to USDS on withdrawal
+ */
+contract SavingUsdsTokenWrapper is BaseTokenWrapper {
+  /**
+   * @dev Constructor
+   * @param tokenIn Address for USDS
+   * @param tokenOut Address for SUSDS
+   * @param pool The address of the Aave Pool
+   * @param owner The address to transfer ownership to
+   */
+  constructor(
+    address tokenIn,
+    address tokenOut,
+    address pool,
+    address owner
+  ) BaseTokenWrapper(tokenIn, tokenOut, pool, owner) {
+    IERC20(tokenIn).approve(tokenOut, type(uint256).max);
+  }
+
+  /// @inheritdoc BaseTokenWrapper
+  function borrowToken(
+    uint256 amount,
+    address to,
+    uint16 referralCode
+  ) external override {
+    _borrowToken(amount, to, referralCode);
+  }
+
+  /// @inheritdoc BaseTokenWrapper
+  function borrowTokenWithPermit(
+    uint256 amount,
+    address to,
+    uint16 referralCode,
+    uint256 deadline,
+    uint8 permitV,
+    bytes32 permitR,
+    bytes32 permitS
+  ) external override {
+    if (deadline != 0) {
+      address debtToken = IPool(0x87870Bca3F3fD6335C3F4ce8392D69350B4fA4E2)
+        .getReserveData(TOKEN_OUT)
+        .variableDebtTokenAddress;
+
+      ICreditDelegationToken(debtToken).delegationWithSig(
+        msg.sender,
+        address(this),
+        amount,
+        deadline,
+        permitV,
+        permitR,
+        permitS
+      );
+    }
+    _borrowToken(amount, to, referralCode);
+  }
+
+  /// @inheritdoc BaseTokenWrapper
+  function getTokenOutForTokenIn(
+    uint256 amount
+  ) external view override returns (uint256) {
+    return IUSDS(TOKEN_OUT).previewDeposit(amount);
+  }
+
+  /// @inheritdoc BaseTokenWrapper
+  function getTokenInForTokenOut(
+    uint256 amount
+  ) external view override returns (uint256) {
+    return IUSDS(TOKEN_OUT).previewRedeem(amount);
+  }
+
+  /// @inheritdoc BaseTokenWrapper
+  function _wrapTokenIn(uint256 amount) internal override returns (uint256) {
+    return IUSDS(TOKEN_OUT).deposit(amount, address(this));
+  }
+
+  /// @inheritdoc BaseTokenWrapper
+  function _unwrapTokenOut(uint256 amount) internal override returns (uint256) {
+    return IUSDS(TOKEN_OUT).redeem(amount, address(this), address(this));
+  }
+
+  function _borrowToken(
```

Comment:
this is the base implementation and can go to the base contract

Reply:
Done [1ab43a0](https://github.com/aave/token-wrappers/commit/1ab43a08563f9e8a0bcaa33acf2cefb3d232359f)

Reply:
can we add the logic to call this internal function also to the base contract?
I d put everything in the base contract, except the implementation of function hooks to wrap/unwrap

Reply:
Done [4f5f382](https://github.com/aave/token-wrappers/commit/4f5f382df666cd1faaa83fded0587213d24f17e4)